programs/gamma/src/fees/dynamic_fee.rs

```diff
@@ -199,113 +217,90 @@ impl DynamicFee {
-                // they are already in descending order of block timestamp.
-                break;
-            }
-            let last_observation_index = if observation_with_index.index == 0 {
-                OBSERVATION_NUM - 1
-            } else {
-                observation_with_index.index as usize - 1
-            };
-
-            // if last observation is not valid, skip this observation
-            if observation_state.observations[last_observation_index].block_timestamp == 0 {
-                continue;
-            }
+            let time_delta = next_obs
+                .block_timestamp
+                .saturating_sub(obs.block_timestamp) as u128;
 
-            if observation_state.observations[last_observation_index].block_timestamp
-                > observation_with_index.observation.block_timestamp
-            {
-                // Break if current observation is older than the last observation.
-                break;
+            // Ensure time_delta is positive
+            if time_delta == 0 {
+                continue;
             }
 
-            let cumulative_token_0_price = observation_with_index
-                .observation
-                .cumulative_token_0_price_x32;
-
-            let last_observation = &observation_state.observations[last_observation_index];
-            let last_cumulative_token_0_price = last_observation.cumulative_token_0_price_x32;
-
-            let time_delta = observation_with_index
-                .observation
-                .block_timestamp
-                .saturating_sub(last_observation.block_timestamp)
-                as u128;
-
-            // Since we are using cumulative_token_price we are sure, that current_observation.cumulative_token_0_price > last_observation.cumulative_token_0_price
-            let price = cumulative_token_0_price
-                .checked_sub(last_cumulative_token_0_price)
+            // Calculate price over the interval
+            let price = next_obs
+                .cumulative_token_0_price_x32
+                .checked_sub(obs.cumulative_token_0_price_x32)
                 .ok_or(GammaError::MathOverflow)?
                 .checked_div(time_delta)
                 .ok_or(GammaError::MathOverflow)?;
 
-            // change cumulative
+            // Update min and max prices
             min_price = min_price.min(price);
             max_price = max_price.max(price);
-            count += 1;
 
-            total_price = total_price
-                .checked_add(price)
-                .ok_or(GammaError::MathOverflow)?;
+            // Accumulate weighted prices for TWAP
+            let price_decimal = Decimal::from_u128(price).ok_or(GammaError::MathOverflow)?;
+            let time_delta_decimal =
+                Decimal::from_u128(time_delta).ok_or(GammaError::MathOverflow)?;
+            weighted_price_sum = weighted_price_sum + (price_decimal * time_delta_decimal);
+            total_weight = total_weight + time_delta_decimal;
```

Comment:
I don't see why we have to calculate it in for loop,

can't we directly get `weighted_price_sum` = first_observation.cumulative_token_0_price_x32-last_observation.cumulative_token_0_price_x32 

Does that not work because of circular indexing?